cpp/src/arrow/python/serialize.h

```diff
@@ -62,13 +64,14 @@ struct ARROW_PYTHON_EXPORT SerializedPyObject {
   /// components as Buffer instances with minimal memory allocation
   ///
   /// {
-  ///   'num_tensors': N,
+  ///   'num_tensors': M,
+  ///   'num_sparse_tensors': N,
   ///   'num_buffers': K,
   ///   'data': [Buffer]
   /// }
   ///
   /// Each tensor is written as two buffers, one for the metadata and one for
-  /// the body. Therefore, the number of buffers in 'data' is 2 * N + K + 1,
+  /// the body. Therefore, the number of buffers in 'data' is 2 * M + 2 * N + K + 1,
```

Comment:
@rok A SparseTensorCOO consists of the indices and the data, so it consists of three buffers (the one is its metadata).  Likewise, a SparseTensorCSR consists of the indices, the indptr, and the data, so it consists of four buffers.

Therefore, we should manage the number of COO and CSR separately.

Reply:
Got it! Indeed this would cause the issue we're seeing.

Reply:
@mrkn I've now changed the [`GetSparseTensorMessage`](https://github.com/apache/arrow/blob/f93c547d6706e7dd2fb77ab5d558bc71fb15f17e/cpp/src/arrow/ipc/writer.cc#L829) so that it concatenates buffers into one and returns a message with two buffers. Do you think this is a valid solution? [`ConcatenateBuffers`](https://github.com/apache/arrow/blob/f93c547d6706e7dd2fb77ab5d558bc71fb15f17e/cpp/src/arrow/buffer.cc#L236) seems to copy data.